src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Models/Catalog/DemoItemEntity.cs

```diff
@@ -1,11 +1,45 @@
 using System.Collections.ObjectModel;
+using System.Linq;
+using VirtoCommerce.CatalogModule.Core.Model;
 using VirtoCommerce.CatalogModule.Data.Model;
+using VirtoCommerce.DemoSolutionFeaturesModule.Core.Models.Catalog;
 using VirtoCommerce.Platform.Core.Common;
 
 namespace VirtoCommerce.DemoSolutionFeaturesModule.Data.Models.Catalog
 {
     public class DemoItemEntity : ItemEntity
     {
         public virtual ObservableCollection<DemoProductPartEntity> ConfiguredProductParts { get; set; } = new NullCollection<DemoProductPartEntity>();
+
+        public override CatalogProduct ToModel(CatalogProduct product, bool convertChildrens = true, bool convertAssociations = true)
+        {
+            var demoProduct = (DemoProduct) base.ToModel(product, convertChildrens, convertAssociations);
```

Comment:
var product = base.ToModel(product, convertChildrens, convertAssociations);

if (product is DemoProduct demoProduct)

Reply:
We override product type so all instances always MUST be of our type. If the system somehow will create an instance of the base type, then it already means it's in an inconsistent state. So I think we shouldn't add additional checks here, just explicitly cast to our type.